### PR TITLE
[MRG] Fixing bug when participants file is a subset of what mne-bids writes

### PR DESCRIPTION
**Pull request by adam2392**
PR Description
--------------

Closes: #623 

Had some time off :).

Merge checklist
---------------

Maintainer, please confirm the following before merging:

- [ ] All comments resolved
- [ ] This is not your own PR
- [ ] All CIs are happy
- [ ] PR title starts with [MRG]
- [ ] [whats_new.rst](https://github.com/mne-tools/mne-bids/blob/master/doc/whats_new.rst) is updated
- [ ] PR description includes phrase "closes <#issue-number>"

### CODE REVIEW

**Comment by adam2392** [comment]
@dnacombo lmk if this fixes it?

**Review by agramfort — APPROVED** [review]
LGTM

**Comment by agramfort** [comment]
@hoechenberger @sappelhoff @jasmainak green button is yours

**Comment by hoechenberger** [comment]
Looking good, but according to codecov, the the `data_value is None` case is not being covered by any tests:
https://codecov.io/gh/mne-tools/mne-bids/compare/409dd36fa1ce422b30b599e26adb09c771818d8d...b34d648e82d2dd28721a066aa258797a019260e7/diff

Could you add a test for this case?

File: mne_bids/tests/test_write.py [hunk 1]
```diff
@@ -121,6 +121,45 @@ def _test_anonymize(raw, bids_path, events_fname=None, event_id=None):
     return bids_root
 
 
+def test_write_participants():
+    """Test that participants data is written correctly."""
```
Comment (jasmainak) [inline mne_bids/tests/test_write.py:125]: aren't there tests for this elsewhere already.

Quickly looked and found this: https://github.com/mne-tools/mne-bids/blob/409dd36fa1ce422b30b599e26adb09c771818d8d/mne_bids/tests/test_write.py#L383-L396

should we be consolidating these?

Comment (hoechenberger) [reply mne_bids/tests/test_write.py:125]: I think it's okay to have the new test function, but maybe change its name and description to reflect that it's specifically meant to check for the issue you're fixing here. And these lines should go:

```python
    # assert age of participant is correct
     participants_tsv = op.join(bids_root, 'participants.tsv')
     data = _from_tsv(participants_tsv)
     assert data['age'][data['participant_id'].index('sub-01')] == '1'
```

Comment (hoechenberger) [reply mne_bids/tests/test_write.py:125]: @jasmainak But I agree with you that we should try to rework, consolidate, clean up, and declutter our tests in general…

**Comment by agramfort** [comment]
my metric of success on tests is : test as well or better than before but
reduce testing time

>

**Comment by adam2392** [comment]
> Looking good, but according to codecov, the the `data_value is None` case is not being covered by any tests:
> https://codecov.io/gh/mne-tools/mne-bids/compare/409dd36fa1ce422b30b599e26adb09c771818d8d...b34d648e82d2dd28721a066aa258797a019260e7/diff
> 
> Could you add a test for this case?

It turns out that line wasn't even needed. TIL:

```Python
data = np.array(None)

data is None
>>> False
```

However, the later code that does masking will return False for the `mask`.
```Python
    mask = None
    for key, row_value in row_data.items():
        # if any of the columns don't even exist in the keys
        # this data_value will return False
        data_value = np.array(data.get(key))

        # Cast row_value to the same dtype as data_value to avoid a NumPy
        # FutureWarning, see
        # https://github.com/mne-tools/mne-bids/pull/372
        row_value = np.array(row_value, dtype=data_value.dtype)

        column_mask = np.in1d(data_value, row_value)
        mask = column_mask if mask is None else (mask & column_mask)
    return np.any(mask)
```

File: mne_bids/tests/test_write.py (hunk 1 above)
Comment (adam2392) [reply mne_bids/tests/test_write.py:125]: I put the participants unit testing code from awhile back into this new unit test function to consolidate testing for participants.tsv file.

File: mne_bids/tests/test_write.py [hunk 2]
```diff
@@ -121,6 +121,95 @@ def _test_anonymize(raw, bids_path, events_fname=None, event_id=None):
...
+    bids_path.update(subject='02')
+    write_raw_bids(raw, bids_path, verbose=False)
+    data = _from_tsv(participants_tsv)
+
+    # hand should have been written properly with now 'n/a' for sub-01
+    # but 'L' for sub-02
+    assert data['hand'][data['participant_id'].index('sub-01')] == 'n/a'
+    assert data['hand'][data['participant_id'].index('sub-02')] == 'L'
+
+    # check to make sure participant data is overwritten, but keeps the fields
```
Comment (adam2392) [inline mne_bids/tests/test_write.py:168]: this and down is copy/pasted from `test_write_correct _inputs`

File: doc/whats_new.rst [hunk 3]
```diff
@@ -58,6 +58,7 @@ Bug fixes
 - Fix a bug in :func:`mne_bids.write_raw_bids`, where original format of data was not kept when writing to FIFF, by `Alexandre Gramfort`_, `Stefan Appelhoff`_, and `Richard Höchenberger`_ (:gh:`610`)
 - Fix a bug where conversion to BrainVision format was done even when non-Volt channel types were present in the data (BrainVision conversion is done by ``pybv``, which currently only supports Volt channel types), by `Stefan Appelhoff`_ (:gh:`619`)
 - Ensure sidecar files (`.tsv` and `.json`) are always read and written in UTF-8, by `Richard Höchenberger`_ (:gh:`625`)
+- Fix a bug where participants files were not being appended to correctly when they contained a subset of ``hand``, ``age`` and ``sex``, by `Adam Li`_ (:gh:`648`)
```
Comment (jasmainak) [inline doc/whats_new.rst:61]: participants files or file? I thought there was only one?

Comment (adam2392) [inline doc/whats_new.rst:61]:
```suggestion
- Fix a bug where participants.tsv file was not being appended to correctly when they contained a subset of ``hand``, ``age`` and ``sex``, by `Adam Li`_ (:gh:`648`)
```

File: doc/whats_new.rst [hunk 4]
```diff
@@ -58,6 +58,7 @@ Bug fixes
 - Fix a bug in :func:`mne_bids.write_raw_bids`, where original format of data was not kept when writing to FIFF, by `Alexandre Gramfort`_, `Stefan Appelhoff`_, and `Richard Höchenberger`_ (:gh:`610`)
 - Fix a bug where conversion to BrainVision format was done even when non-Volt channel types were present in the data (BrainVision conversion is done by ``pybv``, which currently only supports Volt channel types), by `Stefan Appelhoff`_ (:gh:`619`)
 - Ensure sidecar files (`.tsv` and `.json`) are always read and written in UTF-8, by `Richard Höchenberger`_ (:gh:`625`)
+- Fix a bug where participants.tsv file was not being appended to correctly when they contained a subset of ``hand``, ``age`` and ``sex``, by `Adam Li`_ (:gh:`648`)
```
Comment (hoechenberger) [inline doc/whats_new.rst:61]:
```suggestion
- Fix a bug where ``participants.tsv`` was not being appended to correctly when it contained a subset of ``hand``, ``age`` and ``sex``, by `Adam Li`_ (:gh:`648`)
```

**Review by hoechenberger — APPROVED** [review]

**Comment by hoechenberger** [comment]
Thanks, @adam2392!

**Pull request merged**